### PR TITLE
refactor!: Promote SourceLink into a standalone type

### PR DESCRIPTION
**Pull request by paulgessinger**
This promotes the `SourceLink` from an implicit definition + concept
into an actual base class. Source links from experiments (and our tests
& examples) are expected to derive from it, and initialize it with a
`GeometryIdentifier`. This commit also makes sure that the source links
are not copied anymore, since they are only referred to as pointers from
now on.

On major benefit of this change is that `GainMatrixUpdater` and `GainMatrixSmoother` can be untemplated, and therefore also be moved to their own compilation units. Since they are quite heavy, this should give noticeable improvements to build performance for the (C)KF tests and examples.

BREAKING CHANGE: The variant `Measurement` types, `MultiTrajectory`, `GainMatrixUpdater`, `GainMatrixSmoother`, `KalmanFitter`, `CombinatorialKalmanFilter` and some other components lose their `source_link_t` template parameter. `KalmanFitter` also gets a new `fit()` signature, where sourcelinks are given as a pair of iterators.

### CODE REVIEW

File: Core/include/Acts/TrackFitting/detail/VoidKalmanComponents.hpp [hunk 1]
```diff
@@ -45,10 +46,12 @@ struct VoidKalmanUpdater {
   ///
   /// @return The copied predicted parameters
   template <typename track_state_t, typename predicted_state_t>
-  auto operator()(track_state_t& trackState,
-                  const predicted_state_t& predicted) const {
+  Acts::Result<void> operator()(const Acts::GeometryContext& /*gctx*/,
```
Comment (asalzburger) [inline Core/include/Acts/TrackFitting/detail/VoidKalmanComponents.hpp:49]: Why did the return type change by this here?

**Review by asalzburger — APPROVED** [review]
Although this touches many files, it is actually pretty straight forward, once the change to promote to a class is done, the rest is practically direct consequences.

File: Core/include/Acts/TrackFitting/detail/VoidKalmanComponents.hpp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/TrackFitting/detail/VoidKalmanComponents.hpp:49]: Before it was actually just implicit, it returned whatever it returned, and it was the updated parameters, but I think this was either compiled nowhere, or never checked so far.

The `GainMatrixUpdater` already returned `Result<void>` so I'm effectively only harmonizing the interface.

File: Core/src/TrackFitting/CMakeLists.txt [hunk 2]
```diff
@@ -2,4 +2,6 @@ target_sources(
   ActsCore
   PRIVATE
     KalmanFitterError.cpp
+    GainMatrixUpdater.cpp
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/src/TrackFitting/CMakeLists.txt:5]: 👍

**Comment by asalzburger** [comment]
Good to go, after updating ...

**Pull request merged**